### PR TITLE
fix(hesai_hw_interface): change to synchronous requests to prevent undefined sensor behavior

### PR DESCRIPTION
**Pull request by mojomex**
## PR Type

- Bug Fix

## Related Links

[TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-5613) -- Issue
[TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/x/JwRxtQ) -- Investigation

## Description

OT128 had 2s of packets dropped per minute, which was caused by multiple TCP streams being executed concurrently.
Fixing this issue required to
* remove `tcp_driver_s_` and local TCP drivers in `hesai_hw_interface.cpp`
* change from asynchronous requests and nested callbacks to blocking/synchronous-style communication

In the wake of this big refactoring, I made the following other changes:
* remove superfluous function overloads in HW interface
* move PTC header handling and TCP send/receive logic to its own function in HW interface, accessed by all Get/Set functions

## Review Procedure

Refer to the investigation above.
Things I checked (on 40P, QT128, OT128):
* TCP streams not concurrent
* No UDP packet drops
* Pointclouds are correct and at 10Hz without drops
* PTP works as expected

## Remarks

Functionality in `#ifdef WITH_DEBUG_STDOUT_HESAI_HW_INTERFACE` etc. has not been tested.

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp [hunk 1]
```diff
@@ -80,7 +83,7 @@ class Pandar128E4X : public HesaiSensor<hesai_packet::Packet128E4X>
 
     int channel_offset_ns;
     bool is_hires_mode = packet.tail.operational_state == OperationalState::HIGH_RESOLUTION;
-    auto azimuth_state = packet.tail.geAzimuthState(block_id);
+        auto azimuth_state = packet.tail.geAzimuthState(block_id);
```
Comment (drwnz) [inline nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp:86]: Unneccesary tab?

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp [hunk 2]
```diff
@@ -91,7 +94,7 @@ class Pandar128E4X : public HesaiSensor<hesai_packet::Packet128E4X>
         channel_offset_ns = firing_time_offset_as1_ns_[channel_id];
       }
     }
-
+
```
Comment (drwnz) [inline nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp:97]: Unnecessary newline?

File: nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_hesai/hesai_hw_interface.hpp [hunk 3]
```diff
@@ -77,9 +77,12 @@ const uint16_t PANDAR128_E4X_EXTENDED_PACKET_SIZE = 1117;
...
 
-const int PTP_LOG_ANNOUNCE_INTERVAL = 1; // Time interval between Announce messages, in units of log seconds (default: 1)
-const int PTP_SYNC_INTERVAL = 1; //Time interval between Sync messages, in units of log seconds (default: 1)
-const int PTP_LOG_MIN_DELAY_INTERVAL = 0; //Minimum permitted mean time between Delay_Req messages, in units of log seconds (default: 0)
+const int PTP_LOG_ANNOUNCE_INTERVAL =
+  1;  // Time interval between Announce messages, in units of log seconds (default: 1)
+const int PTP_SYNC_INTERVAL =
+  1;  // Time interval between Sync messages, in units of log seconds (default: 1)
+const int PTP_LOG_MIN_DELAY_INTERVAL = 0;  // Minimum permitted mean time between Delay_Req
+                                           // messages, in units of log seconds (default: 0)
```
Comment (drwnz) [inline nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_hesai/hesai_hw_interface.hpp:85]: (pedantic) Inline comments causing a line break

File: nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_hesai/hesai_hw_interface.hpp [hunk 4]
```diff
@@ -916,27 +332,15 @@ class HesaiHwInterface : NebulaHwInterfaceBase
...
-    int profile,
-    int domain,
-    int network,
-    int logAnnounceInterval,
-    int logSyncInterval,
+    std::shared_ptr<boost::asio::io_context> ctx, int profile, int domain, int network,
+    int logAnnounceInterval, int logSyncInterval, int logMinDelayReqInterval);
+  HesaiStatus SetPtpConfigSyncHttp(
+    int profile, int domain, int network, int logAnnounceInterval, int logSyncInterval,
     int logMinDelayReqInterval);
```
Comment (drwnz) [inline nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_hesai/hesai_hw_interface.hpp:339]: (pedantic) this should also be on the previous line

File: nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_hesai/hesai_hw_interface.hpp (hunk 3 above)
Comment (mojomex) [reply nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_hesai/hesai_hw_interface.hpp]: Addressed in [adf43fe](https://github.com/tier4/nebula/pull/126/commits/adf43fef5af5924675b60e317a6de3b59e7a9f06)

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp (hunk 2 above)
Comment (mojomex) [reply nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp]: Addressed in [adf43fe](https://github.com/tier4/nebula/pull/126/commits/adf43fef5af5924675b60e317a6de3b59e7a9f06)

File: nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp (hunk 1 above)
Comment (mojomex) [reply nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/pandar_128e4x.hpp]: Addressed in [adf43fe](https://github.com/tier4/nebula/pull/126/commits/adf43fef5af5924675b60e317a6de3b59e7a9f06)

File: nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp [hunk 5]
```diff
@@ -2924,7 +1154,8 @@ HesaiStatus HesaiHwInterface::CheckAndSetConfig(
 #endif
   auto current_return_mode = nebula::drivers::ReturnModeFromIntHesai(
     hesai_config.return_mode, sensor_configuration->sensor_model);
-  auto wait_time = 100ms; // Avoids spamming the sensor, which leads to failure when configuring it.
+  auto wait_time =
+    100ms;  // Avoids spamming the sensor, which leads to failure when configuring it.
```
Comment (drwnz) [inline nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp:1132]: (pedantic)

File: nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp [hunk 6]
```diff
@@ -3130,78 +1362,30 @@ HesaiStatus HesaiHwInterface::CheckAndSetConfig()
...
-      false);
-    if (stopped) tcp_driver_->run();
-  }
+    std::stringstream ss;
+    ss << result;
+    PrintInfo(ss.str());
+    CheckAndSetConfig(
+      std::static_pointer_cast<HesaiSensorConfiguration>(sensor_configuration_), result);
+  });
+  t.join();  // here
```
Comment (drwnz) [inline nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp:1348]: (pedantic) here?

**Review by drwnz — APPROVED** [review]
LGTM!

File: nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp (hunk 5 above)
Comment (mojomex) [reply nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp]: Addressed in [d56b77e](https://github.com/tier4/nebula/pull/126/commits/d56b77e2adffe2c41f8990cfa4926608cb163fa6)

File: nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp (hunk 6 above)
Comment (mojomex) [reply nebula_hw_interfaces/src/nebula_hesai_hw_interfaces/hesai_hw_interface.cpp]: Addressed in [d56b77e](https://github.com/tier4/nebula/pull/126/commits/d56b77e2adffe2c41f8990cfa4926608cb163fa6)

**Pull request merged**